x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/changepoint/ChangePointAggregator.java

```diff
@@ -0,0 +1,394 @@
+        for (int cp : candidateChangePoints) {
+            double maybeVAlt = (cp * lowerRange.variance() + (n - cp) * upperRange.variance()) / n;
+            if (maybeVAlt < vAlt) {
+                vAlt = maybeVAlt;
+                changePoint = cp;
+            }
+            lowerRange.addValues(timeWindow, cp, cp + step);
+            upperRange.removeValues(timeWindow, cp, cp + step);
+        }
+        dfAlt = n - 2;
+
+        pValueVsNull = independentTrialsPValue(fTestPValue(vNull, dfNull, vAlt, dfAlt), candidateChangePoints.length);
+        if (pValueVsNull < pValueThreshold) {
+            changeType = new ChangeType.StepChange(pValueVsNull, bucketValues.getBucketIndex(changePoint));
+            vNull = vAlt;
+        }
+
+        VarianceAndRValue vAndR = new VarianceAndRValue(Double.MAX_VALUE, Double.MAX_VALUE);
+        changePoint = candidateChangePoints[candidateChangePoints.length - 1] + 1;
+        lowerRange = new RunningStats();
+        upperRange = new RunningStats();
+        // Initialize running stats so that they are only missing the individual changepoint values
+        upperRange.addValues(timeWindow, candidateChangePoints[0], timeWindow.length);
+        lowerRange.addValues(timeWindow, 0, candidateChangePoints[0]);
+        LeastSquaresOnlineRegression lowerLeastSquares = new LeastSquaresOnlineRegression(2);
+        LeastSquaresOnlineRegression upperLeastSquares = new LeastSquaresOnlineRegression(2);
+        for (int i = 0; i < candidateChangePoints[0]; i++) {
+            lowerLeastSquares.add(i, timeWindow[i]);
+        }
+        for (int i = candidateChangePoints[0], x = 0; i < timeWindow.length; i++, x++) {
+            upperLeastSquares.add(x, timeWindow[i]);
+        }
+        // TODO This is crazy inefficient, not only does it do multiple array copies, it calculates r_value without
+        // taking account previous values.
+        double[] monotonicX = new double[timeWindow.length];
+        for (int i = 0; i < timeWindow.length; i++) {
+            monotonicX[i] = i;
+        }
+        int upperMovingWindow = 0;
+        for (int cp : candidateChangePoints) {
+            double lowerRangeVar = lowerRange.variance();
+            double upperRangeVar = upperRange.variance();
+            double rv1 = lowerLeastSquares.squareResidual(
+                monotonicX,
+                0,
+                cp,
+                timeWindow,
+                0,
+                cp,
+                lowerLeastSquares.parameters(),
+                lowerRangeVar
+            );
+            double rv2 = upperLeastSquares.squareResidual(
+                monotonicX,
+                upperMovingWindow,
+                timeWindow.length - cp,
+                timeWindow,
+                cp,
+                timeWindow.length - cp,
+                upperLeastSquares.parameters(),
+                upperRangeVar
+            );
```

Comment:
This is the biggest change in calculation with the online least squares. Since we can only add/remove single values, this recalculates the squared residual at each step. Where each step removes a number of values from the "front" of the input.

Previously, the `timeWindow` values were always paired with `[0..len(time_windows)]` (and their squares). But now its `[iteration_step..len(time_window)]` (and their squares).

I ran this locally and all our currently known cases still pass.

@tveasey ^ What do you think?

Reply:
Yes this works. The online/incremental formulation allows us to both add and remove values, in effect one can compute `R([m, n)) = R([0, n)) - R([0,m))` with `R` denoting least squares regression fit to the data in the interval `[x,y)`.

Note that this formulation does end up squaring the condition number of the matrix you invert, so you have roughly the square root of the precision you can achieve for regression parameters. (There would be ways to mitigate this, but they aren't straightforward.) I don't think this is a problem in practice because these matrices are small, we're working to double precision and we don't really need super high accuracy to detect significant trend changes.